torch/distributed/optim/named_optimizer.py

```diff
@@ -144,14 +144,22 @@ def state_dict(self) -> Dict[str, Any]:
 
         return self._post_state_dict({"state": ret_state, "param_groups": ret_groups})
 
-    def step(self, closure: Any = None) -> None:
+    @overload
+    def step(self, closure: None = ...) -> None:
+        ...
+
+    @overload
```

Comment:
Could you explain why these two overloads aren't already included in the actual step signature below?

Reply:
Only overload signatures are used for signature matching. The last signature without the `@overload` is purely for type hinting.

Reply:
The 2 different signatures say this:

1. The `ellipsis` means that the default argument should be matched i.e. `.step()`. Overall this signature says that if the closure is `None`, i.e. `.step()`, `.step(closure=None)`, or `.step(None)`, then we know that `None` will be returned.
```python
@overload
    def step(self, closure: None = ...) -> None:
        ...
```
2. If a `Callable` that accepts no positional or keyword arguments and returns a `float` is provided as the closure, i.e. `.step(closure=lambda: 0.0)` or `.step(lambda: 0.0)`, then we know a `float` will be returned.
```python
    @overload
    def step(self, closure: Callable[[], float]) -> float:
        ...
```

Reply:
Thanks so much for the exhaustive explanation!